tensorflow/python/distribute/input_lib.py

```diff
@@ -124,24 +126,31 @@ def get_distributed_datasets_from_function(dataset_fn,
         `worker_device_pairs`.
     strategy: a `tf.distribute.Strategy` object, used to run all-reduce to
         handle last partial batch.
+    replication_mode: Replication mode for the input function.
 
   Returns:
     A distributed dataset instance.
   """
   if tf2.enabled():
-    return DistributedDatasetsFromFunction(
-        dataset_fn,
-        input_workers,
-        input_contexts,
-        strategy)
+    if replication_mode == InputReplicationMode.PER_WORKER:
+      return DistributedDatasetsFromFunction(
+          dataset_fn,
+          input_workers,
+          input_contexts,
+          strategy)
+    else:
+      return DistributedDatasetsFromFunctionForReplicas(
```

Comment:
i've carefully checked the code changes in this file, and what I understand is that with the input_workers already adjusted in MirroredStrategy, the only change you need in this file for PER_REPLICA case that is different from PER_WORKER case is that you don't want to use the `(Owned)MultiDeviceIterator` and instead want to do simply `iter(..)`. 
I think all the rest of the code for `DistributedDatasetsFromFunction` should just work as-is. 

So instead of forking all these methods here, I think all you need to do is use the `replication_mode` and `experimental_prefetch_to_device` bits to decide whether to use `(Owned)MultiDeviceIterator` or not. And that should be a much smaller change than the current approach. (Also it would be worth asserting that `len(compute_devices)` == 1 in these cases). 

If this makes sense to you, let's change this code. If not, let's schedule a meeting to go over it. 

Also, if you go this route, most of my current comments in this file become irrelevant. 

ps. The other difference I see, but which I believe should not be there, is that you place the dataset and iterator creation on `with device(GPU:0)`. That is not generally feasible except for DALI case, so I think we had discussed in the past that this implementation here should only place the dataset on the CPU, but the DALI code will take care of adding the `with device(GPU:0)` part.

Reply:
I tried to mitigate this to the _create_iterators_per_replica_with_input_context level of iterator creation. For now I have combined the DistributedDatasetsFromFunctionForReplicas and DistributedDatasetsFromFunction logic but wonder why you assume for per_replica case the the dataset should only be placed on CPU?

Reply:
> 
> 
> @kushanam Can you please resolve conflicts? Thanks!

Conflicts resolved.